File: src/algorithms/sorting/exchange/pancake-sort/sources/pancake-sort.py
def pancake_sort(input_array: list[int]) -> list[int]:  # @step:initialize
    sorted_array = input_array.copy()  # @step:initialize
    array_length = len(sorted_array)  # @step:initialize

    for unsorted_size in range(array_length, 1, -1):
        # Find the index of the maximum element in the unsorted portion
        # @step:find-max
        max_index = 0  # @step:find-max
        for search_index in range(1, unsorted_size):  # @step:compare
            if sorted_array[search_index] > sorted_array[max_index]:  # @step:compare
                max_index = search_index  # @step:compare

        # If the max is not already at the end, flip it there
        if max_index != unsorted_size - 1:
            # Flip max to front if not already there
            if max_index != 0:
                # @step:flip
                flip_left = 0  # @step:flip
                flip_right = max_index  # @step:flip
                while flip_left < flip_right:
                    # @step:swap
                    temporary_value = sorted_array[flip_left]  # @step:swap
                    sorted_array[flip_left] = sorted_array[flip_right]  # @step:swap
                    sorted_array[flip_right] = temporary_value  # @step:swap
                    flip_left += 1
                    flip_right -= 1

            # Flip front to end of unsorted portion
            # @step:flip
            flip_left = 0  # @step:flip
            flip_right = unsorted_size - 1  # @step:flip
            while flip_left < flip_right:
                # @step:swap
                temporary_value = sorted_array[flip_left]  # @step:swap
                sorted_array[flip_left] = sorted_array[flip_right]  # @step:swap
                sorted_array[flip_right] = temporary_value  # @step:swap
                flip_left += 1
                flip_right -= 1

        # The element at unsorted_size - 1 is now in its sorted position
        # @step:mark-sorted

    return sorted_array  # @step:complete

File: src/algorithms/sorting/exchange/pancake-sort/sources/PancakeSort.java
public class PancakeSort {
    public static int[] pancakeSort(int[] inputArray) { // @step:initialize
        int[] sortedArray = inputArray.clone(); // @step:initialize
        int arrayLength = sortedArray.length; // @step:initialize

        for (int unsortedSize = arrayLength; unsortedSize > 1; unsortedSize--) {
            // Find the index of the maximum element in the unsorted portion
            // @step:find-max
            int maxIndex = 0; // @step:find-max
            for (int searchIndex = 1; searchIndex < unsortedSize; searchIndex++) { // @step:compare
                if (sortedArray[searchIndex] > sortedArray[maxIndex]) { // @step:compare
                    maxIndex = searchIndex; // @step:compare
                }
            }

            // If the max is not already at the end, flip it there
            if (maxIndex != unsortedSize - 1) {
                // Flip max to front if not already there
                if (maxIndex != 0) {
                    // @step:flip
                    int flipLeft = 0; // @step:flip
                    int flipRight = maxIndex; // @step:flip
                    while (flipLeft < flipRight) {
                        // @step:swap
                        int temporaryValue = sortedArray[flipLeft]; // @step:swap
                        sortedArray[flipLeft] = sortedArray[flipRight]; // @step:swap
                        sortedArray[flipRight] = temporaryValue; // @step:swap
                        flipLeft++;
                        flipRight--;
                    }
                }

                // Flip front to end of unsorted portion
                // @step:flip
                int flipLeft = 0; // @step:flip
                int flipRight = unsortedSize - 1; // @step:flip
                while (flipLeft < flipRight) {
                    // @step:swap
                    int temporaryValue = sortedArray[flipLeft]; // @step:swap
                    sortedArray[flipLeft] = sortedArray[flipRight]; // @step:swap
                    sortedArray[flipRight] = temporaryValue; // @step:swap
                    flipLeft++;
                    flipRight--;
                }
            }

            // The element at unsortedSize - 1 is now in its sorted position
            // @step:mark-sorted
        }

        return sortedArray; // @step:complete
    }
}
